### DISCUSSION TITLE
Installation on Cray systems ....

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1077] [General]
Hello,

I am trying to compile OpenSCYL on a cray system using the cray compilers. The `cmake` step seems be completed without problem:

``` 
> cmake -DCMAKE_INSTALL_PREFIX=/project/project_465000536/HS/HIPSYCL -DROCM_PATH=$ROCM_PATH -DWITH_CPU_BACKEND=ON -DWITH_CUDA_BACKEND=OFF  -DWITH_ROCM_BACKEND=ON  -DDEFAULT_GPU_ARCH=gfx90a -DWITH_ACCELERATED_CPU=ON ..
-- The C compiler identification is Clang 14.0.6
-- The CXX compiler identification is Clang 14.0.6
-- Cray Programming Environment 2.7.17 C
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/cray/pe/craype/2.7.17/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Cray Programming Environment 2.7.17 CXX
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/cray/pe/craype/2.7.17/bin/CC - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /appl/lumi/SW/LUMI-22.08/G/EB/Boost/1.79.0-cpeCray-22.08/lib64/cmake/Boost-1.79.0/BoostConfig.cmake (found version "1.79.0") found components: context fiber 
-- hip::amdhip64 is SHARED_LIBRARY
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS - Failed
-- Performing Test Terminfo_LINKABLE
-- Performing Test Terminfo_LINKABLE - Success
-- Found Terminfo: /appl/lumi/SW/LUMI-22.08/G/EB/ncurses/6.2-cpeCray-22.08/lib/libtinfo.so  
-- Found ZLIB: /appl/lumi/SW/LUMI-22.08/G/EB/zlib/1.2.12-cpeCray-22.08/lib/libz.so (found version "1.2.12") 
-- Building hipSYCL against LLVM configured from /appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/llvm/lib/cmake/llvm
-- Selecting clang: /appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/llvm/bin/clang++
-- AMD clang version: 5.3.3
-- Using clang include directory: /appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/llvm/lib/clang/15.0.0/include/..
-- Looking for C++ include filesystem
-- Looking for C++ include filesystem - found
-- Performing Test CXX_FILESYSTEM_NO_LINK_NEEDED
-- Performing Test CXX_FILESYSTEM_NO_LINK_NEEDED - Success
-- Found hipRTC: /appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/lib/libhiprtc.so
-- Found Threads: TRUE  
-- Found OpenMP_C: -fopenmp (found version "5.0") 
-- Found OpenMP_CXX: -fopenmp (found version "5.0") 
-- Found OpenMP: TRUE (found version "5.0")  
-- Configuring done
-- Generating done
-- Build files have been written to: /project/project_465000536/HS/OpenSYCL/build
```  

When I compile with `make -j 64` it fails at:

```
[ 81%] Built target opensycl-info
[ 82%] Linking CXX shared library libllvm-to-backend.so
ld.lld: error: unable to find library -lLLVM
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/compiler/llvm-to-backend/CMakeFiles/llvm-to-backend.dir/build.make:129: src/compiler/llvm-to-backend/libllvm-to-backend.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:497: src/compiler/llvm-to-backend/CMakeFiles/llvm-to-backend.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

What is it supposed to be done about `-lLLVM` . The system has many clang compilers, where is the `libLLVM` located?

**Comment by illuhad** [comment]
`libLLVM.so` should be shipped as part of your LLVM installation that you build Open SYCL against. In your case, this is `/appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/llvm`. Please check whether this library exists and whether the linker is given the appropriate directory using `-L`.
Some LLVM installations do not ship the shared library. The SSCP compiler currently requires this however: https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/install-llvm.md#only-if-you-wish-to-compile-llvm-from-source-not-recommended

**Comment by csccva** [comment]
Thanks for the reply. I will try to find the libllvm. It must be there since a EB recipe using the cpeGnu works. Is there some cmake option like `-DLLVM_PATH=...´ ?

**Reply by illuhad** [reply]
There is `-DLLVM_DIR`, but from the output of `cmake` you can already see that it has selected one LLVM: `-- Building hipSYCL against LLVM configured from /appl/lumi/SW/LUMI-22.08/G/EB/rocm/5.3.3/llvm/lib/cmake/llvm`

**Comment by csccva** [comment]
It seems that the llvm library found is not compatible with the cray compilers and this is why is generating the.  So I guess I have to compile it myself using the cray compilers and link to it.

**Comment by csccva** [comment]
I tried the option `-DLLVM_BUILD_LLVM_DYLIB=ON` 

I got this:

```

CMake Warning:
  Manually-specified variables were not used by the project:

    LLVM_BUILD_LLVM_DYLIB
    
    
```



**EDIT** I guess the option is for the llvm compilation not for the hipsycl (facepalm)

**LATER EDIT** I digged through the rocm installation and there is no `libLLVM.so` file.

**Reply by illuhad** [reply]
> EDIT I guess the option is for the llvm compilation not for the hipsycl (facepalm)

Yes.

> LATER EDIT I digged through the rocm installation and there is no libLLVM.so file.

You will either have to compile LLVM from source, or disable hipSYCL's generic single-pass compiler (`-DWITH_SSCP_COMPILER=OFF`). Note that building against ROCm LLVM overall is a bit sketchy, and not recommended because ROCm LLVM does not have a stable API that we can develop against.

**Reply by csccva** [reply]
I will try this. But what does it mean to disable this option? I am trying to prepare a SYCL course on the hope that eventually SYCL will be supported on LUMI later. ( I am not good at installations :( , just trying to get a working compiler and be able to show the most important features)

**Reply by csccva** [reply]
It seems that the installation worked. I tried a simple code with 8 gpus and some device properties query and a simple vector add. I ran the vector add  with default q selector both with 1 GPU and with 1 core (on an node without GPU), so I guess it works. 
I feel like a monkey pressing random buttons  and creating a poem :)

Cristian

**Reply by illuhad** [reply]
>  But what does it mean to disable this option?

It disables support for our generic single-pass compiler (`--hipsycl-targets=generic`). This is arguably the most advanced compilation flow available in any SYCL implementation today. In a single compiler pass, it generates the host binary and embeds the device code in a backend-independent format. Device code is then translated at runtime to whatever is needed: ptx, amdgcn, SPIR-V etc. This allows for faster compile times, while automatically generating a binary that runs "everywhere" (Intel, NVIDIA, AMD GPUs).
https://github.com/OpenSYCL/OpenSYCL/blob/develop/doc/compilation.md#generic-sscp-compilation-flow
https://dl.acm.org/doi/abs/10.1145/3585341.3585351

**Reply by csccva** [reply]
Thanks for the reply. I am more interested in code portability  so I did not give any thought to back-end independent binary.   I am happy with installing it LUMI being able to use it without problems.

Cristian
